src/test/java/com/google/devtools/build/lib/buildeventservice/BuildEventServiceGrpcClientTest.java

```diff
@@ -59,54 +67,92 @@ public void onError(Throwable t) {}
         public void onCompleted() {}
       };
 
-  @Before
-  public void setUp() throws IOException {
-    MockitoAnnotations.initMocks(this);
-    String uniqueName = UUID.randomUUID().toString();
-    InProcessServerBuilder.forName(uniqueName)
-        .directExecutor()
-        .addService(fakeServer)
-        .build()
-        .start();
+  @FunctionalInterface
+  private static interface TestWithServer {
+    void apply(ManagedChannel channel) throws Exception;
+  }
 
+  /**
+   * Test helper that sets up a in-process test server and invokes a callback with a client channel
+   * once the server is running.
+   */
+  private static void withServer(ServerServiceDefinition service, TestWithServer callback)
+      throws Exception {
+    String uniqueName = UUID.randomUUID().toString();
+    Server server =
+        InProcessServerBuilder.forName(uniqueName).directExecutor().addService(service).build();
+    server.start();
     ManagedChannel channel = InProcessChannelBuilder.forName(uniqueName).directExecutor().build();
-    grpcClient = new BuildEventServiceGrpcClient(channel, null);
+    try {
```

Comment:
Test helpers which encapsulate the control flow of the entire test case tend to obscure the actual test behavior, which can frustrate investigation when tests break. In this case, the value of the helper method is abstracting the construction and teardown of the server. We can define a helper for that, without requiring the caller provide a callback.

The static helper can return an instance of an `AutoCloseable` class which includes the `#shutdown()` calls in its `#close()` method. Then the test cases can be written as:

```
@Test
public void besHeaders() throws Exception {
    ArrayList<Metadata> seenHeaders = new ArrayList<>();
    try (TestChannelAndServer channelAndServer = startInprocessServer(ServerInterceptors.intercept(...))) {
         BuildEventServiceGrpcClient grpcClient =
              new BuildEventServiceGrpcClient(
                  channel, null, ImmutableList.of(Maps.immutableEntry("metadata-foo", "bar")));
          assertThat(grpcClient.openStream(ack -> {}).getStatus().get()).isEqualTo(Status.OK);
          assertThat(seenHeaders).hasSize(1);
          Metadata headers = seenHeaders.get(0);
          assertThat(headers.get(Metadata.Key.of("metadata-foo", Metadata.ASCII_STRING_MARSHALLER)))
              .isEqualTo("bar");
    }
}
```

Reply:
Thanks—I like this suggestion a lot.